File: backend/app/schemas/reports.py
"""Audit & compliance report response schemas (Issue #51).

Reports are computed on the fly from existing tables — no new persistence
model is introduced. See ``app.api.v1.reports`` for the aggregation logic.
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

ViolationType = Literal["naming_non_compliant", "rejected"]


class NamingViolationItem(BaseModel):
    container_id: str
    identifier: str
    title: str
    violation_type: ViolationType
    reason: str | None
    occurred_at: str | None
    current_state: str
    current_assignee_id: str | None


class NamingViolationsResponse(BaseModel):
    items: list[NamingViolationItem]
    total: int


class ApprovalDelayAssignee(BaseModel):
    assignee_id: str
    task_type: str
    status: str


class ApprovalDelayItem(BaseModel):
    workflow_id: str
    workflow_type: str
    target_type: str
    target_id: str
    container_identifier: str | None
    container_title: str | None
    created_at: datetime
    elapsed_hours: float
    assignees: list[ApprovalDelayAssignee]


class ApprovalDelaysResponse(BaseModel):
    items: list[ApprovalDelayItem]
    total: int
    threshold_hours: float


class RequirementsStatusItem(BaseModel):
    document_id: str
    doc_type: str
    title: str
    revision: str
    met_count: int
    partial_count: int
    not_met_count: int
    total_count: int
    fulfillment_rate: float


class RequirementsStatusResponse(BaseModel):
    items: list[RequirementsStatusItem]
    total: int

File: frontend/src/api/reports.ts
import { api } from "@/lib/api";

export type ViolationType = "naming_non_compliant" | "rejected";

export interface NamingViolationItem {
  container_id: string;
  identifier: string;
  title: string;
  violation_type: ViolationType;
  reason: string | null;
  occurred_at: string | null;
  current_state: string;
  current_assignee_id: string | null;
}

export interface NamingViolationsResponse {
  items: NamingViolationItem[];
  total: number;
}

export interface ApprovalDelayAssignee {
  assignee_id: string;
  task_type: string;
  status: string;
}

export interface ApprovalDelayItem {
  workflow_id: string;
  workflow_type: string;
  target_type: string;
  target_id: string;
  container_identifier: string | null;
  container_title: string | null;
  created_at: string;
  elapsed_hours: number;
  assignees: ApprovalDelayAssignee[];
}

export interface ApprovalDelaysResponse {
  items: ApprovalDelayItem[];
  total: number;
  threshold_hours: number;
}

export interface RequirementsStatusItem {
  document_id: string;
  doc_type: string;
  title: string;
  revision: string;
  met_count: number;
  partial_count: number;
  not_met_count: number;
  total_count: number;
  fulfillment_rate: number;
}

export interface RequirementsStatusResponse {
  items: RequirementsStatusItem[];
  total: number;
}

export const reportsApi = {
  getNamingViolations: (projectId: string): Promise<NamingViolationsResponse> =>
    api
      .get<NamingViolationsResponse>(
        `/projects/${projectId}/reports/naming-violations`,
      )
      .then((r) => r.data),

  getApprovalDelays: (
    projectId: string,
    thresholdHours?: number,
  ): Promise<ApprovalDelaysResponse> =>
    api
      .get<ApprovalDelaysResponse>(
        `/projects/${projectId}/reports/approval-delays`,
        thresholdHours !== undefined
          ? { params: { threshold_hours: thresholdHours } }
          : undefined,
      )
      .then((r) => r.data),

  getRequirementsStatus: (
    projectId: string,
  ): Promise<RequirementsStatusResponse> =>
    api
      .get<RequirementsStatusResponse>(
        `/projects/${projectId}/reports/requirements-status`,
      )
      .then((r) => r.data),
};
